File: medium/disjoint_set.py
class DisjointSet:
    def __init__(self, size: int):
        self.root = [i for i in range(size)]
        self.rank = [0] * size
        self.setCount = size

    def find(self, node: int) -> int:
        if self.root[node] != node:
            self.root[node] = self.find(self.root[node])
        return self.root[node]

    def union(self, node1: int, node2: int):
        root1 = self.find(node1)
        root2 = self.find(node2)
        if root1 == root2:
            return
        if self.rank[root1] > self.rank[root2]:
            self.root[root2] = root1
        elif self.rank[root1] < self.rank[root2]:
            self.root[root1] = root2
        else:
            self.rank[root1] += 1
            self.root[root2] = root1
        self.setCount -= 1
    
    def getSetCount(self) -> int:
        return self.setCount



def makeDS(edges: List[List[int]]) -> int:
    n = len(edges)
    ds = DisjointSet(n)
    for edge in edges:
        ds.union(edge[0], edge[1])

    setCount = ds.getSetCount()
    return setCount

File: medium/disjoint_set.go
package main

type UnionFind struct {
    root     []int
    rank     []int
    setCount int
}

func DisjointSet(size int) *UnionFind {
    root := make([]int, size)
    rank := make([]int, size)
    for i := 0; i < size; i++ {
        root[i] = i
        rank[i] = 1
    }
    return &UnionFind{root, rank, size}
}

func (ds *UnionFind) Find(node int) int {
    if node != ds.root[node] {
        ds.root[node] = ds.Find(ds.root[node])
    }
    return ds.root[node]
}

func (ds *UnionFind) Union(node1, node2 int) {
    root1 := ds.Find(node1)
    root2 := ds.Find(node2)
    if root1 == root2 {
        return
    }
    if ds.rank[root1] > ds.rank[root2] {
        ds.root[root2] = root1
    } else if ds.rank[root1] < ds.rank[root2] {
        ds.root[root1] = root2
    } else {
        ds.root[root1] = root2
        ds.rank[root2]++
    }
    ds.setCount--
}

func (ds *UnionFind) getSetCount() int {
    return ds.setCount
}

func makeDS(edges [][]int) int {
    n := len(edges)
    ds := DisjointSet(n)
    for _, edge := range edges {
        ds.Union(edge[0], edge[1])
    }

    return ds.getSetCount()
}
